test/cri-containerd/clone_test.go

```diff
@@ -0,0 +1,387 @@
+// +build functional
+
+package cri_containerd
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"strings"
+	"testing"
+	"time"
+
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+// returns a request config for creating a template sandbox
+// Name field is empty so it must be filled before actually sending this request.
+func getTemplatePodConfig() *runtime.RunPodSandboxRequest {
+	return &runtime.RunPodSandboxRequest{
+		Config: &runtime.PodSandboxConfig{
+			Metadata: &runtime.PodSandboxMetadata{
+				Name:      "",
+				Uid:       "0",
+				Namespace: testNamespace,
+			},
+			Annotations: map[string]string{
+				"io.microsoft.virtualmachine.saveastemplate": "true",
+			},
+		},
+		RuntimeHandler: wcowHypervisorRuntimeHandler,
+	}
+}
+
+// returns a request config for creating a template container
+// Name field is empty so it must be filled before actually sending this request.
+func getTemplateContainerConfig() *runtime.CreateContainerRequest {
+	return &runtime.CreateContainerRequest{
+		Config: &runtime.ContainerConfig{
+			Metadata: &runtime.ContainerMetadata{
+				Name: "",
+			},
+			Image: &runtime.ImageSpec{
+				Image: imageWindowsNanoserver,
+			},
+			// Do not keep the ping running on template containers.
+			Command: []string{
+				"cmd",
+				"/c",
+				"ping",
+				"127.0.0.1",
+			},
+			Annotations: map[string]string{
+				"io.microsoft.virtualmachine.saveastemplate": "true",
+			},
+		},
+	}
+}
+
+// returns a request config for creating a standard container
+// Name field is empty so it must be filled before actually sending this request.
+func getStandardContainerConfig() *runtime.CreateContainerRequest {
+	return &runtime.CreateContainerRequest{
+		Config: &runtime.ContainerConfig{
+			Metadata: &runtime.ContainerMetadata{
+				Name: "",
+			},
+			Image: &runtime.ImageSpec{
+				Image: imageWindowsNanoserver,
+			},
+			Command: []string{
+				"cmd",
+				"/c",
+				"ping",
+				"-t",
+				"127.0.0.1",
+			},
+		},
+	}
+}
+
+// returns a create cloned sandbox request config. This config can not be used
+// as it is because `Name` and `Annotation: templateid` fields are empty
+func getClonedPodConfig() *runtime.RunPodSandboxRequest {
```

Comment:
Why not pass in the fields that we end up editing from the caller into this function?

Reply:
Good point. I have fixed it.